### PR TITLE
 support for the JSON.TOGGLE command

### PR DESCRIPTION
**Pull request by MaheshMadushan**
[Add support for the JSON.TOGGLE command #1825]

### CODE REVIEW

File: src/types/json.h [hunk 1]
```diff
@@ -173,6 +173,22 @@ struct JsonValue {
     return types;
   }
 
+  Status Toggle(std::string_view path) {
+    try {
+      jsoncons::jsonpath::json_replace(value, path, [&](const std::string & /*path*/, jsoncons::json &val) {
+        bool is_boolean = val.is_bool() && !val.empty();
```
Comment (PragmaTwice) [inline src/types/json.h]: May I ask why we check `empty()` here? Seems `is_bool()` is enough to me.

Comment (MaheshMadushan) [reply src/types/json.h]: Thanks for pointing out. fixed

File: src/types/json.h [hunk 2]
```diff
@@ -173,6 +173,22 @@ struct JsonValue {
...
 
+  Status Toggle(std::string_view path) {
+    try {
+      jsoncons::jsonpath::json_replace(value, path, [&](const std::string & /*path*/, jsoncons::json &val) {
+        bool is_boolean = val.is_bool();
+
+        if (is_boolean)
+          val = !val.as_bool();
+        else
+          return;
```
Comment (PragmaTwice) [inline src/types/json.h]:
```suggestion
        if (val.is_bool()) val = !val.as_bool();
```

Comment (MaheshMadushan) [reply src/types/json.h]: Thanks for the suggestion. fixed it.

File: src/commands/cmd_json.cc [hunk 3]
```diff
@@ -189,6 +189,24 @@ class CommandJsonClear : public Commander {
...
+ public:
+  Status Execute(Server *svr, Connection *conn, std::string *output) override {
+    redis::Json json(svr->storage, conn->GetNamespace());
+
+    std::string path = (args_.size() > 2) ? args_[2] : "$";
+    auto s = json.Toggle(args_[1], path);
+
+    if (s.IsNotFound()) {
+      return Status::OK();
+    }
```
Comment (PragmaTwice) [inline src/commands/cmd_json.cc]: Have you tested this case? I think here the connection will be blocked.

Comment (MaheshMadushan) [reply src/commands/cmd_json.cc]: Not yet. Writing more test cases for happy path. Will update asap with error scenarios.

Comment (PragmaTwice) [reply src/commands/cmd_json.cc]: You can add `*output = redis::NilString()` before returning to solve it, just like in other commands.

Comment (MaheshMadushan) [reply src/commands/cmd_json.cc]: Sure will do

**Review by PragmaTwice — DISMISSED** [review]

**Comment by PragmaTwice** [comment]
You can try to build it to make sure it will pass the compiler before pushing.

**Comment by MaheshMadushan** [comment]
> You can try to build it to make sure it will pass the compiler before pushing.

issue fixed.

File: src/types/json.h [hunk 4]
```diff
@@ -211,6 +211,17 @@ struct JsonValue {
     return types;
   }
 
+  Status Toggle(std::string_view path) {
+    try {
+      jsoncons::jsonpath::json_replace(value, path, [&](const std::string & /*path*/, jsoncons::json &val) {
+        if (val.is_bool()) val = !val.as_bool();
```
Comment (git-hulk) [inline src/types/json.h]: > JSON.TOGGLE returns an array of integer replies for each path, the new value (0 if false or 1 if true), or nil for JSON values matching the path that are not Boolean. 

The return value should be an array of integers. That said if there are two paths matched and one of them is NOT a boolean value, then it should return an array with:

1) 0(or 1)
2) nil

Can refer to other command implementations.

Comment (MaheshMadushan) [reply src/types/json.h]: Thanks for poiniting out. will fix this

Comment (git-hulk) [reply src/types/json.h]: Thank you!

File: src/types/json.h [hunk 5]
```diff
@@ -211,6 +211,23 @@ struct JsonValue {
     return types;
   }
 
+  StatusOr<std::vector<std::optional<int>>> Toggle(std::string_view path) {
+    std::vector<std::optional<int>> result;
```
Comment (PragmaTwice) [inline src/types/json.h]: Seems here we can just use a `std::vector<std::optional<bool>>`?

Comment (MaheshMadushan) [reply src/types/json.h]: Thanks for pointing out. That is more suitable. Fixed.

File: src/commands/cmd_json.cc [hunk 6]
```diff
@@ -219,6 +219,35 @@ class CommandJsonClear : public Commander {
...
+
+    if (s.IsNotFound()) {
+      *output = redis::NilString();
+      return Status::OK();
+    }
+
+    *output = redis::MultiLen(results.size());
+    for (auto it = results.rbegin(); it != results.rend(); ++it) {
+      if ((*it).has_value()) {
+        *output += redis::Integer((*it).value());
```
Comment (PragmaTwice) [inline src/commands/cmd_json.cc]:
```suggestion
      if (it->has_value()) {
        *output += redis::Integer(it->value());
```

Comment (MaheshMadushan) [reply src/commands/cmd_json.cc]: Thanks for poiniting out. fixed

**Review by git-hulk — APPROVED** [review]
LGTM, thanks

**Comment by MaheshMadushan** [comment]
> LGTM, thanks

Thanks for the feedback

**Comment by git-hulk** [comment]
@MaheshMadushan Thanks for your contribution.

**Pull request merged**